### PR TITLE
Use `parallel_hash` join algorithm by default

### PR DESCRIPTION
**Pull request by nickitat**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Now `parallel_hash` algorithm will be used (if applicable) when `join_algorithm` setting is set to `default`. Two previous alternatives (`direct` and `hash`) are still considered when `parallel_hash` cannot be used.

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit e2c34e0012f647cb129ac87835388294aebcea89 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/70788/e2c34e0012f647cb129ac87835388294aebcea89/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/70768/b69ce95a134ea98daa1d430f338adcb48c65a899/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Comment by nickitat** [comment]
makes sense also to choose CHJ in `JoinSwitcher` (i.e. for `join_algorithm='auto'`).

**Comment by nickitat** [comment]
extracted fix to backport in release branch: https://github.com/ClickHouse/ClickHouse/pull/72405

**Review by vdimir — APPROVED** [review]

File: src/Core/Settings.cpp [hunk 1]
```diff
@@ -2461,7 +2461,7 @@ Possible values:
 
 - default
 
- This is the equivalent of `hash` or `direct`, if possible (same as `direct,hash`)
+ This is the equivalent of `hash`, `parallel_hash` or `direct`, if possible (same as `direct,parallel_hash,hash`)
```
Comment (rschu1ze) [inline src/Core/Settings.cpp:2464]: The join algorithms in the first part of the sentence are in the wrong order.

Maybe just

```
Same as `direct,parallel_hash,hash`, i.e. try using `direct`, `parallel_hash` or `hash` in this order and if applicable each.
```

?

**Comment by rschu1ze** [comment]
@qoega `parallel hash` should (in theory) be strictly better than `hash`. Can you point to queries that became slower or fail? If they fail, is it because of functional bugs or because of OOM?

Technically, this PR didn't change setting `join_algorithm`. It only changed the meaning of setting value `default`. We would need a new setting value `default_new` (or some similarly awkwardly named setting value) to move `default` to compat settings.

**Comment by qoega** [comment]
Shared examples

**Pull request merged**